File: android_xml/app/src/main/java/com/example/androidxml/features/todo/TodoListViewModel.kt
package com.example.androidxml.features.todo

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.map

data class TodoListState(
    val todos: List<Todo> = dummyTodos,
    val filter: TodoFilter = TodoFilter.ALL,
    val isLoading: Boolean = false,
    val isSyncing: Boolean = false,
    val error: String? = null,
)

class TodoListViewModel : ViewModel() {
    private val _state = MutableLiveData(TodoListState())

    val filtered: LiveData<List<Todo>> = _state.map { state ->
        when (state.filter) {
            TodoFilter.ALL       -> state.todos
            TodoFilter.ACTIVE    -> state.todos.filter { !it.isCompleted }
            TodoFilter.COMPLETED -> state.todos.filter { it.isCompleted }
        }
    }

    val filter: LiveData<TodoFilter> = _state.map { it.filter }

    fun setFilter(filter: TodoFilter) {
        _state.value = _state.value?.copy(filter = filter)
    }

    fun toggle(id: String) {
        val todos = _state.value?.todos ?: return
        _state.value = _state.value?.copy(
            todos = todos.map { if (it.id == id) it.copy(isCompleted = !it.isCompleted) else it }
        )
    }

    fun delete(id: String) {
        val todos = _state.value?.todos ?: return
        _state.value = _state.value?.copy(todos = todos.filter { it.id != id })
    }
}

File: ios_swiftui/TodoApp/Features/Todo/TodoListViewModel.swift
import Foundation

// Flutter の TodoListNotifier に相当
@Observable
@MainActor
final class TodoListViewModel {
    var todos: [Todo] = []
    var filter: TodoFilter = .all
    var isLoading = true
    var errorMessage: String? = nil
    var isSyncing = false

    private let repository: TodoRepositoryInterface

    init(repository: TodoRepositoryInterface) {
        self.repository = repository
    }

    var filteredTodos: [Todo] {
        switch filter {
        case .all: todos
        case .active: todos.filter { !$0.isCompleted }
        case .completed: todos.filter { $0.isCompleted }
        }
    }

    func load() async {
        isLoading = true
        defer { isLoading = false }
        do {
            todos = try await repository.load()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func toggle(id: String) async {
        guard let i = todos.firstIndex(where: { $0.id == id }) else { return }
        todos[i].isCompleted.toggle()
        do {
            try await repository.update(todos[i])
        } catch {
            todos[i].isCompleted.toggle()
            errorMessage = error.localizedDescription
        }
    }

    func delete(id: String) async {
        todos.removeAll { $0.id == id }
        do {
            try await repository.delete(id: id)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func saveTodo(_ todo: Todo) async {
        if todos.contains(where: { $0.id == todo.id }) {
            todos.removeAll { $0.id == todo.id }
            todos.insert(todo, at: 0)
            do { try await repository.update(todo) }
            catch { errorMessage = error.localizedDescription }
        } else {
            todos.insert(todo, at: 0)
            do { try await repository.save(todo) }
            catch { errorMessage = error.localizedDescription }
        }
    }

    func syncFromAPI() async {
        isSyncing = true
        defer { isSyncing = false }
        do {
            let added = try await repository.sync()
            todos.append(contentsOf: added)
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}
